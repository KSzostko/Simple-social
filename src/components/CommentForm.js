import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { connect } from 'react-redux';
import { Formik, Form, Field, ErrorMessage } from 'formik';
import * as Yup from 'yup';
import { addComment } from '../redux/comment/commentActions'
import Button from './Button';

const StyledForm = styled(Form)`
    display: flex;
    flex-direction: column;
    align-items: center;
`;

const StyledHeader = styled.h2`
    margin-top: 0;
    margin-bottom: 40px;
`;

const StyledWrapper = styled.div`
    width: 100%;
    margin-bottom: 20px;
    display: grid;
    grid-template-columns: auto 1fr;
`;

const StyledLabel = styled.label`
    margin-top: 10px;
    margin-right: 40px;
    font-size: 12px;
`;

const StyledInputField = styled(Field)`
    margin-left: 0;
    padding: 5px 10px;
    border: 1px solid #BBB;
    border-radius: 4px;

    &:focus {
        border-color: #252B2D;
    }
`;

const StyledTextareaField = styled(Field)`
    resize: none;
    padding: 5px 10px;
    border: 1px solid #BBB;
    border-radius: 4px;

    &:focus {
        border-color: #252B2D;
    }
`;

const StyledButtonWrapper = styled.div`
    margin-top: 20px;
    width: 100%;
    display: flex;
    justify-content: flex-end;
`;

const StyledButton = styled(Button)`
    margin-right: 20px;
`;

function CommentForm(props) {
    const { closeModal, postId , addCommentFn} = props;
    
    const initialValues = {
        name: '',
        email: '',
        body: '',
    };

    const validationSchema = Yup.object({
        name: Yup.string().required('Required'),
        email: Yup.string().email('Invalid email format').required('Required'),
        body: Yup.string().required('Required'),
    });

    const onSubmit = values => {
        values.postId = postId;

        closeModal();

        addCommentFn(values);
    };
    
    return (
        <Formik
            initialValues={initialValues}
            validationSchema={validationSchema}
            onSubmit={onSubmit}
        >
            {formik => (
                <StyledForm>
                    <StyledHeader>Add comment</StyledHeader>
                    <StyledWrapper>
                        <StyledLabel htmlFor="name">Name</StyledLabel>
                        <StyledInputField
                            id="name" 
                            name="name"
                        />
                        <ErrorMessage name="name" />
                    </StyledWrapper>
                    <StyledWrapper>
                        <StyledLabel htmlFor="email">Email</StyledLabel>
                        <StyledInputField
                            type="email"
                            id="email" 
                            name="email"
                        />
                        <ErrorMessage name="email" />
                    </StyledWrapper>
                    <StyledWrapper>
                        <StyledLabel htmlFor="body">Body</StyledLabel>
                        <StyledTextareaField
                            component="textarea"
                            id="body"
                            name="body"
                            rows="6"
                        />
                        <ErrorMessage name="body" />
                    </StyledWrapper>
                    <StyledButtonWrapper>
                        <StyledButton type="button" onClick={closeModal}>Cancel</StyledButton>
                        <Button type="submit" disabled={!formik.isValid}>Save</Button>
                    </StyledButtonWrapper>
                </StyledForm>
            )}
        </Formik>
    );
}

const mapDispatchToProps = dispatch => {
    return {
        addCommentFn: comment => dispatch(addComment(comment)),
    };
}

CommentForm.propTypes = {
    closeModal: PropTypes.func,
    addCommentFn: PropTypes.func,
    postId: PropTypes.string,
};

export default connect(null, mapDispatchToProps)(CommentForm);