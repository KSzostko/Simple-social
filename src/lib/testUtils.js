const fakePost = () => ({
    userId: 1,
    postId: 2,
    title: 'Post title',
    body: 'Post body',
});

const fakeComment = () => ({
    postId: 1,
    id: 1,
    name: 'Comment name',
    email: 'user@gmail.com',
    body: 'Comment body',
});

export { fakePost, fakeComment };