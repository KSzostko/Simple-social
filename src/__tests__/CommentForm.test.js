import React from 'react';
import { render, screen, fireEvent, wait } from '@testing-library/react';
import mockAxios from 'axios';
import { Provider } from 'react-redux';
import store from '../redux/store';
import { fakeComment } from '../lib/testUtils';
import CommentForm from '../components/CommentForm';

jest.mock('axios');

const fakeData = {
    closeModal: () => {},
    postId: '1',
};

describe('<CommentForm />', () => {
    it('renders and matches the snapshot', () => {
        const { container } = render(
            <Provider store={store}>
                <CommentForm {...fakeData} />
            </Provider>
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it('will not submit form if it is not fully filled out, it displays errors instead and disables the submit button', async () => {
        const fakeCommentData = fakeComment();
        
        render(
            <Provider store={store}>
                <CommentForm {...fakeData} />
            </Provider>
        );

        await wait(() => fireEvent.click(screen.getByText(/save/i)));
        expect(screen.getByText(/save/i)).toBeDisabled();

        await wait(() => fireEvent.change(screen.getByLabelText(/name/i), {
            target: {
                value: fakeCommentData.name,
            }
        }));
        await wait(() => fireEvent.change(screen.getByLabelText(/body/i), {
            target: {
                value: fakeCommentData.body,
            }
        }));
        expect(screen.getByText(/required/i)).toBeInTheDocument();
        expect(mockAxios.post).not.toHaveBeenCalled();
    });

    it('will not submit form if email format is not right', async () => {
        render(
            <Provider store={store}>
                <CommentForm {...fakeData} />
            </Provider>
        );

        await wait(() => fireEvent.click(screen.getByText(/save/i)));
        expect(screen.getByText(/save/i)).toBeDisabled();

        await wait(() => fireEvent.change(screen.getByLabelText(/email/i), {
            target: {
                value: 'email',
            }
        }));
        expect(screen.getByText(/invalid email format/i)).toBeInTheDocument();
        expect(mockAxios.post).not.toHaveBeenCalled();
    });

    it('will submit form if it is fully filled out', async () => {
        const fakePostData = fakeComment();

        mockAxios.post.mockResolvedValue();
        
        render(
            <Provider store={store}>
                <CommentForm {...fakeData} />
            </Provider>
        );

        await wait(() => fireEvent.change(screen.getByLabelText(/name/i), {
            target: {
                value: fakePostData.name,
            }
        }));
        await wait(() => fireEvent.change(screen.getByLabelText(/body/i), {
            target: {
                value: fakePostData.body,
            }
        }));
        await wait(() => fireEvent.change(screen.getByLabelText(/email/i), {
            target: {
                value: fakePostData.email,
            }
        }));
        expect(screen.queryByText(/required/i)).not.toBeInTheDocument();
        expect(screen.getByText(/save/i)).not.toBeDisabled();
        
        await wait(() => fireEvent.click(screen.getByText(/save/i)));
        expect(mockAxios.post).toHaveBeenCalledTimes(1);
    });
});