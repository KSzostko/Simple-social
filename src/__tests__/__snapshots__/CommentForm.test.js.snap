// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CommentForm /> renders and matches the snapshot 1`] = `
<form
  action="#"
  class="sc-AxirZ cHkTwW"
>
  <h2
    class="sc-AxiKw gvpAQo"
  >
    Add comment
  </h2>
  <div
    class="sc-AxhCb WMzQB"
  >
    <label
      class="sc-AxhUy glMlo"
      for="name"
    >
      Name
    </label>
    <input
      class="sc-fzozJi lkIOAO"
      id="name"
      name="name"
      value=""
    />
  </div>
  <div
    class="sc-AxhCb WMzQB"
  >
    <label
      class="sc-AxhUy glMlo"
      for="email"
    >
      Email
    </label>
    <input
      class="sc-fzozJi lkIOAO"
      id="email"
      name="email"
      type="email"
      value=""
    />
  </div>
  <div
    class="sc-AxhCb WMzQB"
  >
    <label
      class="sc-AxhUy glMlo"
      for="body"
    >
      Body
    </label>
    <textarea
      class="sc-AxgMl khZarX"
      id="body"
      name="body"
      rows="6"
    />
  </div>
  <div
    class="sc-AxheI ecSaBX"
  >
    <button
      class="sc-AxjAm sc-Axmtr debgrO"
      type="button"
    >
      Cancel
    </button>
    <button
      class="sc-AxjAm clzjYv"
      type="submit"
    >
      Save
    </button>
  </div>
</form>
`;
